llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -739,6 +766,29 @@ bool AIEPostRASchedStrategy::tryCandidate(SchedCandidate &Cand,
       return TryCand.Reason != NoCand;
     }
 
+    // Special heuristics for loops.
+    // Note that they aren't used for the first fixpoint iteration: this is
+    // currently a workaround because we want a very optimistic schedule in that
+    // first iteration. That is because it decides the slot assignments for
+    // multi-slot instructions. This rule can probably be deleted once the
+    // loop-aware scheduler knows how to reassign those.
+    const BlockState &BS = getInterBlock().getBlockState(CurMBB);
+    if (UseLoopHeuristics && BS.Kind == AIE::BlockType::Loop &&
+        BS.getRegions().size() == 1 && BS.FixPoint.NumIters > 0) {
+      const InterBlockEdges &LoopEdges = BS.getBoundaryEdges();
+
+      // For instructions with equal dependence chains, prioritize scheduling
+      // instructions that are used later in the next iteration. The point is
+      // to teach our heuristics a tiny bit about LCDs.
+      if (tryLess(getEarliestLoopCarriedUse(*TryCand.SU, LoopEdges) +
+                      TryCand.SU->BotReadyCycle,
+                  getEarliestLoopCarriedUse(*Cand.SU, LoopEdges) +
+                      Cand.SU->BotReadyCycle,
+                  TryCand, Cand, BotPathReduce)) {
+        return TryCand.Reason != NoCand;
+      }
```

Comment:
AFAICS, this uses scheduling from the previous fixpoint iteration, which seems to me the best rationale not to trust the first scheduling run.

Reply:
We do not really use the scheduling from the first iteration, only the "boundary DDG" and the height of nodes in the "post-boundary" half.

Reply:
Still.  BotReadyCycle sounds like a scheduling artifact to me. Or is this initialized with its height?

Reply:
Oh sorry I mis-interpreted your comment. `getEarliestLoopCarriedUse(*TryCand.SU, LoopEdges)` is constant, but `Cand.SU->BotReadyCycle` obviously is not and will change as the loop schedule is converging. But if we want a number to fully trust, there's only the last iteration that will give it.